### PR TITLE
Scoped the modifiers to convey intended uses (NUnit.Framework.Constraints)

### PR DESCRIPTION
**Pull request by jnm2**
In the same vein as https://github.com/nunit/nunit/pull/2804.

ℹ This is not why I think `sealed` is important, but a side effect of this is that recent JITs generate faster code when calling methods of sealed classes.

### CODE REVIEW

File: src/NUnitFramework/framework/Assume.cs [hunk 1]
```diff
@@ -34,7 +34,7 @@ namespace NUnit.Framework
     /// result. If an assumption is not met, the test
     /// should produce an inconclusive result.
     /// </summary>
-    public class Assume
+    public abstract class Assume
```
Comment (jnm2) [inline src/NUnitFramework/framework/Assume.cs]: This would be `static` (= `sealed abstract`) but we want to allow inheritance in these particular classes as an extensibility mechanism since C# does not allow static extension members.

File: src/NUnitFramework/framework/Attributes/RepeatAttribute.cs [hunk 2]
```diff
@@ -72,7 +72,7 @@ public TestCommand Wrap(TestCommand command)
         /// <summary>
         /// The test command for the RepeatAttribute
         /// </summary>
-        public class RepeatedTestCommand : DelegatingTestCommand
+        private sealed class RepeatedTestCommand : DelegatingTestCommand
```
Comment (jnm2) [inline src/NUnitFramework/framework/Attributes/RepeatAttribute.cs]: This class provides no new surface area. Extending code would use `RepeatAttribute.Wrap` to obtain these instances and use composition to delegate to them.

File: src/NUnitFramework/framework/Compatibility/LongLivedMarshalByRefObject.cs [hunk 3]
```diff
@@ -30,11 +30,11 @@ namespace NUnit.Compatibility
     /// A MarshalByRefObject that lives forever
     /// </summary>
 #if NETSTANDARD1_6
-    public class LongLivedMarshalByRefObject
+    public abstract class LongLivedMarshalByRefObject
     {
     }
 #else
-    public class LongLivedMarshalByRefObject : MarshalByRefObject
+    public abstract class LongLivedMarshalByRefObject : MarshalByRefObject
```
Comment (jnm2) [inline src/NUnitFramework/framework/Compatibility/LongLivedMarshalByRefObject.cs]: ⚠

File: src/NUnitFramework/framework/Constraints/CollectionTally.cs [hunk 4]
```diff
@@ -28,22 +28,22 @@ namespace NUnit.Framework.Constraints
...
-        public class CollectionTallyResult
+        public sealed class CollectionTallyResult
         {
             /// <summary>Items that were not in the expected collection.</summary>
-            public List<object> ExtraItems { get; set; }
+            public List<object> ExtraItems { get; }
 
             /// <summary>Items that were not accounted for in the expected collection.</summary>
-            public List<object> MissingItems { get; set; }
+            public List<object> MissingItems { get; }
```
Comment (jnm2) [inline src/NUnitFramework/framework/Constraints/CollectionTally.cs]: Rather than allowing two ways to mutate MissingItems, I went along with the way we were currently doing it.

File: src/NUnitFramework/framework/Constraints/ConstraintBuilder.cs [hunk 5]
```diff
@@ -34,14 +34,14 @@ namespace NUnit.Framework.Constraints
     /// input constraints as well as the results of each
     /// operator applied.
     /// </summary>
-    public class ConstraintBuilder : IResolveConstraint
+    public sealed class ConstraintBuilder : IResolveConstraint
```
Comment (jnm2) [inline src/NUnitFramework/framework/Constraints/ConstraintBuilder.cs]: As far as I can tell, there is no useful way to inherit from this class. If a class is not specifically designed with inheritance in mind, we should signal that like we do with readonly on fields. Inheritance comes with often-unintended consequences. (A simple example is that straight-forward custom equality comparisons stop being symmetric).

File: src/NUnitFramework/framework/Constraints/ConstraintExpression.cs [hunk 6]
```diff
@@ -45,7 +45,7 @@ public class ConstraintExpression
         /// <summary>
         /// The ConstraintBuilder holding the elements recognized so far
         /// </summary>
-        protected ConstraintBuilder builder;
+        protected readonly ConstraintBuilder builder;
```
Comment (jnm2) [inline src/NUnitFramework/framework/Constraints/ConstraintExpression.cs]: Noticed that I missed this in the `readonly` PR. Maybe I should put this in its own PR?

File: src/NUnitFramework/framework/Constraints/Tolerance.cs [hunk 7]
```diff
@@ -324,20 +324,20 @@ private Range PercentRange(object value)
         /// Tolerance.Range represents the range of values that match
         /// a specific tolerance, when applied to a specific value.
         /// </summary>
-        public class Range
+        public struct Range
```
Comment (jnm2) [inline src/NUnitFramework/framework/Constraints/Tolerance.cs]: Since the purpose is to be a value bag rather than to have identity.

File: src/NUnitFramework/framework/Internal/TestListener.cs [hunk 8]
```diff
@@ -21,46 +21,28 @@
...
 namespace NUnit.Framework.Internal
 {
     /// <summary>
-    /// TestListener provides an implementation of ITestListener that
-    /// does nothing. It is used only through its NULL property.
+    /// Provides an implementation of <see cref="ITestListener"/> through
+    /// its <see cref="NULL"/> property that does nothing.
     /// </summary>
-    public class TestListener : ITestListener
+    public static class TestListener
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/TestListener.cs]: The side effect here is that .ToString() gives `NullTestListener` and you check `== TestListener.NULL` rather than `is TestListener`. I was motivated to do this since `is TestListener` doesn't read as though you are checking for a *null* test listener.

**Comment by ChrisMaddock** [comment]
Eee there's a lot to public classes here! 🙀 

What was your logic as to where to make changes here - are you applying this to everything that 'would' be internal (let's avoid the debate on terminology!), or just everything which isn't extended within the framework itself? Or is it classes that make no logical sense to extend? Is the aim also to restrict what we currently expose, or just the semantics/JIT benefits? 

Main question: how do we know this isn't going to break builds, and if we don't know that, what's the justification that makes it worthwhile? 😄

**Comment by jnm2** [comment]
@ChrisMaddock My motivation was to improve the signal to noise ratio especially when browsing our APIs but also when working in the codebase. That includes signalling our semantics and reducing anything we expose which is not designed for public consumption.

I didn't want to really tackle the visibility aspects much in this PR. I'm thinking that should be done namespace by namespace, or even smaller.
The logic in this PR is *mostly* that only classes that are designed for inheritance are left unsealed. It's documentation letting you know that there wouldn't have been anything useful to gain by inheriting. (Or in the case of `abstract`/`static`, by instantiating.)

**Comment by ChrisMaddock** [comment]
Thanks for clarifying! 🙂 

I completely agree this is where we _want_ to be. My concern is how we get there, causing the least disruption to legacy builds. This sort of incremental change, and picking off the low-hanging fruit may be the right way - I guess the other option is bulk making things private. The only issue is that whenever we start touching things like this, we invariably seem to end up breaking _some_ unusual use-case implemented 10 years ago! Maybe it's more down to us to say 'Sorry - that's unsupported' - at least this will move us in the right direction to need to say that less often!

There's a lot of changes here. I'll try and look through at some point.

File: src/NUnitFramework/framework/Constraints/CollectionTally.cs [hunk 9]
```diff
@@ -28,22 +28,22 @@ namespace NUnit.Framework.Constraints
...
 
             /// <summary>Constructs an empty <see cref="CollectionTallyResult"/>.</summary>
-            public CollectionTallyResult()
+            public CollectionTallyResult(List<object> missingItems, List<object> extraItems)
             {
-                ExtraItems = new List<object>();
-                MissingItems = new List<object>();
+                MissingItems = missingItems;
+                ExtraItems = extraItems;
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Constraints/CollectionTally.cs]: Why this change? And if this change, does the doccoment need updating too - as surely this allows you to potentially create a non-empty CollectionTallyResult

**Comment by ChrisMaddock** [comment]
Suggestion for next time - would you mind breaking wide-spread changes like this up into multiple PRs? That would allow us to review/merge 20 at a time or so - most of my time is available in dribs and drabs, and this is a difficult PR to part-review. Thanks! 😄

**Comment by jnm2** [comment]
> Suggestion for next time - would you mind breaking wide-spread changes like this up into multiple PRs? That would allow us to review/merge 20 at a time or so - most of my time is available in dribs and drabs, and this is a difficult PR to part-review. Thanks! 😄

Good idea. Maybe per namespace? It's not too late for me to break this one up...

File: src/NUnitFramework/framework/Constraints/CollectionTally.cs (hunk 9 above)
Comment (jnm2) [reply src/NUnitFramework/framework/Constraints/CollectionTally.cs]: It allowed
```cs
return new CollectionTallyResult()
{
    MissingItems = new List<object>(_missingItems),
    ExtraItems = new List<object>(_extraItems)
};
```
to become:
```cs
return new CollectionTallyResult(
    new List<object>(_missingItems),
    new List<object>(_extraItems));
```
rather than:
```cs
var temp = new CollectionTallyResult();
temp.MissingItems.AddRange(_missingItems);
temp.ExtraItems.AddRange(_extraItems);
return temp;
```

Thanks for catching the doc comment! What do you think, is this too far? Should we go back to allowing the list instances (themselves mutable) to be swapped out, even though we only used that for initialization? What would it mean to later on set them to null or a new list, as opposed to clearing them?

**Comment by ChrisMaddock** [comment]
If you wouldn't mind - that would be very helpful for me. I don't know most of these type names well enough to provide any sort of useful review just from looking at the diff - and I rarely have time to look up 155 different classes in the codebase! 

@rprouse or @CharliePoole may be happier to review this as is though, based on their familiarity of the codebase however!

**Comment by jnm2** [comment]
I would not mind and I'm happy you brought it up! I'll make this PR just about NUnit.Framework.Constraints since that's where you've commented so far.

**Comment by jnm2** [comment]
> The only issue is that whenever we start touching things like this, we invariably seem to end up breaking some unusual use-case implemented 10 years ago! Maybe it's more down to us to say 'Sorry - that's unsupported' - at least this will move us in the right direction to need to say that less often!

There's part of me that wants the pain to be over sooner rather than later. But that doesn't mean this is the right approach. If anything just bothers you, I'd rather not have it in the PR.

**Comment by jnm2** [comment]
PR is updated! Tell me what you think of the list of commits.
I'm going to wait to submit the others until people have had a chance to digest this one.

File: src/NUnitFramework/framework/Constraints/FloatingPointNumerics.cs [hunk 10]
```diff
@@ -45,9 +44,8 @@ namespace NUnit.Framework.Constraints
     ///     as low as 0.0000001 for small numbers or as high as 10.0 for large numbers.
     ///   </para>
     /// </remarks>
-    public class FloatingPointNumerics
+    internal static class FloatingPointNumerics
```
Comment (jnm2) [inline src/NUnitFramework/framework/Constraints/FloatingPointNumerics.cs]: Due to the inaccessible constructor, this class was already effectively `static` from the outside.

File: src/NUnitFramework/framework/Constraints/ConstraintExpression.cs [hunk 11]
```diff
@@ -66,6 +65,7 @@ public ConstraintExpression()
         /// <param name="builder">The builder.</param>
         public ConstraintExpression(ConstraintBuilder builder)
         {
+            Guard.ArgumentNotNull(builder, nameof(builder));
             this.builder = builder;
```
Comment (jnm2) [inline src/NUnitFramework/framework/Constraints/ConstraintExpression.cs]: All the code in this class assumes that `builder` is not null. Now that the field is readonly and this guard is in place, that is finally a safe assumption.

File: src/NUnitFramework/framework/Constraints/CollectionTally.cs (hunk 9 above)
Comment (ChrisMaddock) [reply src/NUnitFramework/framework/Constraints/CollectionTally.cs]: I understand, thanks. Sounds sensible - maybe just the comment to change. 😄

File: src/NUnitFramework/framework/Constraints/ConstraintBuilder.cs (hunk 5 above)
Comment (ChrisMaddock) [reply src/NUnitFramework/framework/Constraints/ConstraintBuilder.cs]: Sounds ok to me

File: src/NUnitFramework/framework/Constraints/ConstraintBuilder.cs [hunk 12]
```diff
@@ -96,7 +96,7 @@ public ConstraintOperator Pop()
         /// <summary>
         /// ConstraintStack is a type-safe stack for holding Constraints
         /// </summary>
-        public class ConstraintStack
+        public sealed class ConstraintStack
```
Comment (ChrisMaddock) [inline src/NUnitFramework/framework/Constraints/ConstraintBuilder.cs]: Is there a reason to not make this internal in-line with the change to OperatorStack?

File: src/NUnitFramework/framework/Constraints/FloatingPointNumerics.cs (hunk 10 above)
Comment (ChrisMaddock) [reply src/NUnitFramework/framework/Constraints/FloatingPointNumerics.cs]: It looks like lots of this class isn't used by the framework. I'm not sure on it's origins - are the public methods designed to be used by end users, or has internal usage of it just lapsed? 

The former would be a reason not to make it internal, of course. `static` seems fine to me.

File: src/NUnitFramework/framework/Constraints/ConstraintBuilder.cs (hunk 12 above)
Comment (jnm2) [reply src/NUnitFramework/framework/Constraints/ConstraintBuilder.cs]: ConstraintOperator has `public abstract void Reduce(ConstraintBuilder.ConstraintStack stack)`.
My bad, `OperatorStack` could have been `private`. I'll change it.

**Comment by jnm2** [comment]
@ChrisMaddock Forgot to ping you after the updates, sorry.

**Comment by ChrisMaddock** [comment]
Looks good. My only concern is not understanding the origin/purpose of the FloatingPointNumerics class. (It being tested so heavily but unused makes me wonder if it was previously available to users, and legacy test suites may still rely on it.) @rprouse or @CharliePoole may know better there. 🙂

**Comment by CharliePoole** [comment]
`FloatingPointNumerics` is used by `Numerics` and comes into play whenever you test floating point values for equality `.Within(n).Ulps`. That's fairly esoteric but you'd be surprised at how much discussion of it there was when I implemented it initially. The class was never aimed at direct user usage.

**Comment by mikkelbu** [comment]
@jnm2 It looks as though AppVeyor is confused (perhaps due to a force push?). 
image

**Comment by jnm2** [comment]
Thanks! I went to https://ci.appveyor.com/project/CharliePoole/nunit/history, used the browser to find the first occurrence of "2816", opened the build and clicked "Re-build PR." Looks like that has fixed it.
